test/SparqlAntlrParserTestHelpers.h

```diff
@@ -368,121 +442,142 @@ MATCHER_P3(IsSelect, distinct, reduced, selection, "") {
+                     &GraphPatternOperation::GroupGraphPattern::_child,
+                     subMatcher));
 };
 
-auto IsUnion = [](auto&& subMatcher1, auto&& subMatcher2) {
-  return testing::Field(
-      "variant_", &GraphPatternOperation::variant_,
-      testing::VariantWith<GraphPatternOperation::Union>(testing::AllOf(
-          testing::Field("_child1", &GraphPatternOperation::Union::_child1,
-                         subMatcher1),
-          testing::Field("_child2", &GraphPatternOperation::Union::_child2,
-                         subMatcher2))));
+auto IsUnion =
+    [](auto&& subMatcher1,
+       auto&& subMatcher2) -> testing::Matcher<const GraphPatternOperation&> {
+  return IsGraphPatternOperation<GraphPatternOperation::Union>(testing::AllOf(
+      testing::Field("_child1", &GraphPatternOperation::Union::_child1,
+                     subMatcher1),
+      testing::Field("_child2", &GraphPatternOperation::Union::_child2,
+                     subMatcher2)));
 };
 
-auto IsMinus = [](auto&& subMatcher) {
-  return testing::Field(
-      "variant_", &GraphPatternOperation::variant_,
-      testing::VariantWith<GraphPatternOperation::Minus>(testing::Field(
-          "_child", &GraphPatternOperation::Minus::_child, subMatcher)));
+auto IsMinus =
+    [](auto&& subMatcher) -> testing::Matcher<const GraphPatternOperation&> {
+  return IsGraphPatternOperation<GraphPatternOperation::Minus>(testing::Field(
+      "_child", &GraphPatternOperation::Minus::_child, subMatcher));
 };
 
-MATCHER_P3(IsGraphPattern, optional, filters, childMatchers, "") {
-  if (arg._graphPatterns.size() != std::tuple_size_v<decltype(childMatchers)>) {
-    return false;
+// TODO: generalize or rename
+template <auto SubMatcherLambda>
+struct MatcherOverloads {
+  testing::Matcher<const GraphPatternOperation&> operator()(
+      vector<SparqlFilter>&& filters, const auto&... childMatchers) {
+    return SubMatcherLambda(std::move(filters), childMatchers...);
   }
 
-  // TODO<joka921, qup42> I think there is a `tupleForEach` function somewhere
-  //  in `util/xxx.h` that could be used to make this eve more idiomatic.
-  auto lambda = [&](auto&&... matchers) {
-    size_t i = 0;
-    return (... && testing::Value(arg._graphPatterns[i++], matchers));
-  };
-  bool childrenMatch = std::apply(lambda, childMatchers);
+  testing::Matcher<const GraphPatternOperation&> operator()(
+      const auto&... childMatchers) {
+    return SubMatcherLambda({}, childMatchers...);
+  }
+};
 
-  return arg._optional == optional &&
-         testing::Value(arg._filters,
-                        testing::UnorderedElementsAreArray(filters)) &&
-         childrenMatch;
-}
+template <auto SubMatcherLambda>
+struct MatcherOverloadss {
+  testing::Matcher<const ParsedQuery::GraphPattern&> operator()(
+      bool optional, vector<SparqlFilter>&& filters,
+      const auto&... childMatchers) {
+    return SubMatcherLambda(optional, std::move(filters), childMatchers...);
+  }
```

Comment:
I think you yet have to find out, how to make the overload work, but we can talk about this (and better namings)
in the next meeting.